---
title: "DNA methylation in differentially methylated probes and regions"
date: "Created: 2021-04-28 Updated: `r Sys.Date()`"
output: 
  html_notebook:
    codes: hide
---

### Introdution

To investigate whether the methylated gene would affect its transcription, we firstly find the overlaps between two-omics layer data, and then show the relationship via visualization. 


#### Infinium MethylationEPIC Manifest Column Headings


Gene region feature category describing the CpG position, from UCSC.
Features listed in the same order as the target gene transcripts.

<details>
<summary>DNA methylation UCSC_RefGene_Group</summary>


<p align="center">
<img src="./images/gene_structure.jpg" width="800" height="500">
</p>


* **TSS200** = 0–200 bases upstream of the transcriptional start site (TSS)

* **TSS1500** = 200–1500 bases upstream of the TSS

* **5'UTR** = Within the 5' untranslated region, between the TSS and the ATG start site

* **1st exon**

* **Body** = Between the ATG and stop codon; irrespective of the presence of introns, exons, TSS, or
promoters

* **exon boundaries**

* **3'UTR** = Between the stop codon and poly A signal.


**Notes: [promoter regions (including TSS200 or/and TSS1500)](https://doi.org/10.3389/fmolb.2021.597513)** 

</details>


#### Purposes 

DNA methylation, one of the most studied epigenetic chagnes in human cells would interface the expression of target genes, where heavily methylated promoter regions in principle correlate with repressed transcription.

Firstly, We performed differential methylation analysis on Infinium MethylationEPIC array probe sites to identify the DEP in each group. We next summarised the numbers of each gene group. 

we dissected the methylation profiles whose the genic locations from upstream promoter regions to the 3' untranslated regions to integrate the DNA methylation profile and transcriptomic profile.  


```{r setup, include=FALSE}
knitr::opts_chunk$set(warning = F, message = F)
options(warn = -1)
library(dplyr)
library(tibble)
library(ggplot2)
library(convert)
library(data.table)
options(warn = 0)

# rm(list = ls())
options(stringsAsFactors = F)
options(future.globals.maxSize = 1000 * 1024^2)
```


### load data 
```{r}
phen <- read.csv("../../Result/phenotype/Pancreatic_tumor_merge_20210426.csv")

# gene
DEG_PPC <- fread("../../Result/Differential/Paracancerous_PrimaryCancer_limma_Gene.csv")
DEG_PCL <- fread("../../Result/Differential/PrimaryCancer_Liver_limma_Gene.csv")

# methylation DMPs
DMP_PPC <- fread("../../Result/Differential/Paracancerous_PrimaryCancer_DMP.csv")
DMP_PCL <- fread("../../Result/Differential/PrimaryCancer_Liver_DMP.csv")

# methylation profile
prf_PPC <- readRDS("../../Result/profile/Paracancerous_PrimaryCancer_MethExprSet_bVals.RDS")
prf_PCL <- readRDS("../../Result/profile/PrimaryCancer_Liver_MethExprSet_bVals.RDS")

# probe annotation
reannEPIC <- readRDS("../../Result/GeneID/EPIC_reannotation.RDS")
```


### Methylation level of individual EPIC array probes
```{r, fig.width=8, fig.height=6}
# Differential Expressed Position
probes_volcano <- function(datset=DMP_PPC,
                           group_name=c("ParacancerousTissue", "PrimaryCancerTissue"),
                           group_col=c("#397A60", "#FC6940"),
                           pval=0.05, 
                           fc=1){
  
  # datset=DMP_PPC
  # group_name=c("ParacancerousTissue", "PrimaryCancerTissue")
  # group_col=c("#397A60", "#FC6940")
  # pval=0.05
  # fc=1
  
  dat <- datset %>% 
    mutate(color = case_when(logFC > fc & adj.P.Val < pval ~ group_name[1],
                             logFC < -fc & adj.P.Val < pval ~ group_name[2],
                             adj.P.Val > pval | abs(logFC) < fc ~ "Nonsignif")) %>%
    mutate(color=factor(color, levels = c(group_name, "Nonsignif")))

  dat_status <- table(dat$color)
  dat_status_number <- as.numeric(dat_status)
  dat_status_ratio <- paste0(signif(dat_status_number/sum(dat_status_number), 3)*100, "%")
  dat_status_name <- names(dat_status)
  legend_label <- c(paste0(dat_status_name[1], " (", dat_status_number[1], ": ", dat_status_ratio[1], ")"),
                    paste0(dat_status_name[2], " (", dat_status_number[2], ": ", dat_status_ratio[2], ")"),
                    paste0("Nonsignif", " (", dat_status_number[3], ": ", dat_status_ratio[3], ")"))
  
  group_col_new <- c(group_col, "#BFBFBF")
  group_name_new <- levels(dat$color)
  
  xlabel <- paste0("log2(FC) [", paste(group_name, collapse=":"), "]")
  
  # Make a basic ggplot2 object with x-y values
  pl <- ggplot(dat, aes(x = logFC, y = -log10(adj.P.Val), color = color))+ 
          geom_point(size = 0.5, alpha = 0.8)+
          scale_color_manual(name = "hypomethylated loci",
                             values = group_col_new,
                             labels = c(legend_label, "Nonsignif"))+
          labs(title = "Methylation level of individual EPIC array probes", 
               x = xlabel,
               y = expression(-log[10]("adjusted p-value")))+
          geom_hline(yintercept=-log10(pval), alpha=.8, linetype=2, size=.7)+
          geom_vline(xintercept=fc, alpha=.8, linetype=2, size=.7)+
          geom_vline(xintercept=-fc, alpha=.8, linetype=2, size=.7)+ 
          annotate("text", x=min(dat$logFC), y=-log10(pval), label=pval, size=6, color="red")+
          annotate("text", x=fc, y=0, label=fc, size=6, color="red")+
          annotate("text", x=-fc, y=0, label=-fc, size=6, color="red")+
          scale_y_continuous(trans = "log1p")+
          guides(color=guide_legend(override.aes = list(size = 3)))+
          theme_bw()+ 
          theme(plot.title = element_text(color = "black", size = 14, face = "bold", hjust = .5),
                axis.title = element_text(color = "black", size = 12),
                axis.text = element_text(color = "black", size = 10),
                text = element_text(size = 8, color = "black", family="serif"),
                panel.grid = element_blank(),
                legend.position = c(.15, .1),
                legend.key.height = unit(0.6, "cm"),
                legend.text = element_text(face = "bold", color = "black", size = 8))
  return(pl)
}

DMP_volcano_PPC <- probes_volcano(datset=DMP_PPC,
              group_name=c("ParacancerousTissue", "PrimaryCancerTissue"),
              group_col=c("#397A60", "#FC6940"),
              pval=0.05, 
              fc=1)
DMP_volcano_PPC
if(!dir.exists("../../Result/Figure")){
  dir.create("../../Result/Figure")
}
ggsave("../../Result/Figure/Paracancerous_PrimaryCancer_DMP_volcano.pdf", DMP_volcano_PPC, width = 8, height = 6)


DMP_volcano_PCL <- probes_volcano(datset=DMP_PCL,
              group_name=c("PrimaryCancerTissue", "LiverMetastasis"),
              group_col=c("#397A60", "#FC6940"),
              pval=0.05, 
              fc=1)
DMP_volcano_PCL
if(!dir.exists("../../Result/Figure")){
  dir.create("../../Result/Figure")
}
ggsave("../../Result/Figure/PrimaryCancer_Liver_DMP_volcano.pdf", DMP_volcano_PCL, width = 8, height = 6)
```


### Different genome regions

* **TSS200** = 0–200 bases upstream of the transcriptional start site (TSS)

* **TSS1500** = 200–1500 bases upstream of the TSS

* **5'UTR** = Within the 5' untranslated region, between the TSS and the ATG start site

* **Body** = Between the ATG and stop codon; irrespective of the presence of introns, exons, TSS, or
promoters

* **3'UTR** = Between the stop codon and poly A signal.

```{r, fig.width=8, fig.height=8}
# Methylation level by genic region
DMP_violin <- function(profile=prf_PPC,
                       dataDMP=DMP_PPC,
                       group_name=c("ParacancerousTissue", "PrimaryCancerTissue"),
                       group_col=c("#397A60", "#FC6940")){
  
  # profile=prf_PPC
  # dataDMP=DMP_PPC
  # group_name=c("ParacancerousTissue", "PrimaryCancerTissue")
  # group_col=c("#397A60", "#FC6940")
  
  
  # dissecting the genic region and gene names
  dataDMP_temp <- dataDMP %>% #dplyr::slice(1:500) %>%
    filter(Hypomethylated != "Nonsignif") %>%
    dplyr::select(GeneID, UCSC_RefGene_Name, UCSC_RefGene_Group)
  
  dat_Group <- apply(dataDMP_temp, 1, function(x){
    
    # x <- as.matrix(dataDMP_temp[3, ])
    G_gene <- unlist(strsplit(x[2], ";"))
    G_group <- unlist(strsplit(x[3], ";"))
    
    if(length(G_gene) == 0){
      G_gene <- NA
      G_group <- "Intergenic"
      G_name <- x[1]
    }else{
      G_name <- rep(x[1], length(G_gene))      
    }
    res <- cbind(G_name, G_group, G_gene)
    return(res)
  }) 
  
  dat_group <- data.frame(do.call(rbind, dat_Group)) %>%
    filter(G_group != "Intergenic")
  rownames(dat_group) <- NULL
  
  # merge data 
  mdat <- full_join(dat_group, dataDMP %>% 
                       dplyr::select(GeneID, Block, logFC, adj.P.Val,
                                     Hypomethylated, AveExpr, t, P.Value, B, Delta) %>%
                      filter(Hypomethylated != "Nonsignif"),
                     by = c("G_name" = "GeneID")) %>%
    dplyr::rename(GeneID=G_name) %>%
    dplyr::rename(UCSC_RefGene_Group=G_group) %>%
    mutate(Hypomethylated=factor(as.character(Hypomethylated), levels = group_name)) 
  mdat_sig <- mdat %>% filter(Hypomethylated != "Nonsignif") #%>%
    #dplyr::slice(1:1000)
  
  # beta value %methylation
  edata <- data.frame(exprs(profile))
  mdat_bVals <- full_join(mdat_sig , 
                           edata %>% rownames_to_column("GeneID"),
                           by = "GeneID")
  mdat_bVals <- mdat_bVals[!is.na(mdat_bVals$UCSC_RefGene_Group), ]
  
  plot_volinFun <- function(tag){
    
    # tag=group_name[1]
    
    dat_tag <- mdat_bVals %>% filter(Hypomethylated == tag) %>%
      dplyr::select(UCSC_RefGene_Group, starts_with("P0")) %>%
      mutate(UCSC_RefGene_Group=factor(as.character(UCSC_RefGene_Group), 
                  levels = c("TSS1500","TSS200","5\'UTR","1stExon", "Body", "ExonBnd", "3\'UTR"))) 
    # summary the number of genic regions
    dat_status <- table(dat_tag$UCSC_RefGene_Group)
    print(dat_status)
    
    xaxis_lab <- c()
    for(i in 1:length(dat_status)){
      xaxis_lab <- c(xaxis_lab, paste0(names(dat_status)[i], "\n(", as.numeric(dat_status)[i], ")"))
    }
    
    predata <-  dat_tag %>%
      tidyr::gather(key="SampleID", value="bVals", -UCSC_RefGene_Group)
    
    pheno <- pData(profile) %>% dplyr::select(Type) %>%
      rownames_to_column("SampleID")
    
    plotdata <- inner_join(pheno, predata, by="SampleID") %>%
      mutate(Type=factor(as.character(Type), levels = group_name))
    
    # Grouped violin plot with split violins
    GeomSplitViolin <- ggproto("GeomSplitViolin", GeomViolin, 
                               draw_group = function(self, data, ..., draw_quantiles = NULL) {
      data <- transform(data, xminv = x - violinwidth * (x - xmin), xmaxv = x + violinwidth * (xmax - x))
      grp <- data[1, "group"]
      newdata <- plyr::arrange(transform(data, x = if (grp %% 2 == 1) xminv else xmaxv), if (grp %% 2 == 1) y else -y)
      newdata <- rbind(newdata[1, ], newdata, newdata[nrow(newdata), ], newdata[1, ])
      newdata[c(1, nrow(newdata) - 1, nrow(newdata)), "x"] <- round(newdata[1, "x"])
    
      if (length(draw_quantiles) > 0 & !scales::zero_range(range(data$y))) {
        stopifnot(all(draw_quantiles >= 0), all(draw_quantiles <=
          1))
        quantiles <- ggplot2:::create_quantile_segment_frame(data, draw_quantiles)
        aesthetics <- data[rep(1, nrow(quantiles)), setdiff(names(data), c("x", "y")), drop = FALSE]
        aesthetics$alpha <- rep(1, nrow(quantiles))
        both <- cbind(quantiles, aesthetics)
        quantile_grob <- GeomPath$draw_panel(both, ...)
        ggplot2:::ggname("geom_split_violin", grid::grobTree(GeomPolygon$draw_panel(newdata, ...), quantile_grob))
      }
      else {
        ggplot2:::ggname("geom_split_violin", GeomPolygon$draw_panel(newdata, ...))
      }
    })  
    
    geom_split_violin <- function(mapping = NULL, 
                                  data = NULL, 
                                  stat = "ydensity", 
                                  position = "identity", 
                                  ..., 
                                  draw_quantiles = NULL, 
                                  trim = TRUE, 
                                  scale = "area", 
                                  na.rm = FALSE, 
                                  show.legend = NA, 
                                  inherit.aes = TRUE) {
      layer(data = data, 
            mapping = mapping, 
            stat = stat, 
            geom = GeomSplitViolin, 
            position = position, 
            show.legend = show.legend, 
            inherit.aes = inherit.aes, 
            params = list(trim = trim, 
                          scale = scale, 
                          draw_quantiles = draw_quantiles, 
                          na.rm = na.rm, 
                          ...))
    }
   
    pl <- ggplot(plotdata, aes(x=UCSC_RefGene_Group, y=bVals, fill = Type))+ 
            geom_split_violin()+
            ggtitle(paste0("Methylation level by genic region", " (", tag, " genes)"))+
            xlab("")+
            ylab("%Methylation")+
            scale_fill_manual(name = NULL,
                               values = group_col,
                               labels = group_name)+
            scale_y_continuous(labels = scales::percent, limits=c(0, 1.2))+
            scale_x_discrete(breaks=c("TSS1500","TSS200","5\'UTR","1stExon","Body","ExonBnd","3\'UTR"),
                             labels=xaxis_lab)+
            theme_minimal()+
            theme(plot.title = element_text(color = "black", size = 12, face = "bold", hjust = .5),
                  axis.title = element_text(color = "black", size = 10),
                  axis.text = element_text(color = "black", size = 10),
                  text = element_text(size = 8, color = "black", family="serif"),
                  #panel.grid = element_blank(),
                  legend.position = "bottom",
                  legend.key.height = unit(0.6, "cm"),
                  legend.text = element_text(face = "bold", color = "black", size = 8)) 
    return(pl)
  }

  p1 <- plot_volinFun(tag = group_name[1])
  p2 <- plot_volinFun(tag = group_name[2])
  
  res <- list(grp1=p1, grp2=p2)
  
}

if(0){
# old version of genic region
DMP_violin <- function(profile=prf_PPC,
                       dataDMP=DMP_PPC,
                       group_name=c("ParacancerousTissue", "PrimaryCancerTissue"),
                       group_col=c("#397A60", "#FC6940")){
  
  # profile=prf_PPC
  # dataDMP=DMP_PPC
  # group_name=c("ParacancerousTissue", "PrimaryCancerTissue")
  # group_col=c("#397A60", "#FC6940")
  
  require(DEP)
  gene_uniq_name <- make_unique(dataDMP, "UCSC_RefGene_Name", "UCSC_RefGene_Group", delim = ";") %>%
    dplyr::select(GeneID, name, ID) %>%
    filter(!name %in% grep("^.\\d", name, value=T)) %>%
    filter(name != "", ID != "") %>%
    setNames(c("GeneID", "UCSC_RefGene_Name", "UCSC_RefGene_Group")) %>%
    # filter(UCSC_RefGene_Group %in% RefGene_Group) %>%
    mutate(Gene=gsub("\\.\\d+$", "", UCSC_RefGene_Name),
           RefGene=paste0(UCSC_RefGene_Group, "_", Gene))
  
  # merge data 
  mdat <- inner_join(gene_uniq_name, 
                     dataDMP %>% 
                       dplyr::select(GeneID, Block, logFC, adj.P.Val,
                                     Hypomethylated, AveExpr, t, P.Value, B, Delta),
                     by ="GeneID") %>%
    mutate(Hypomethylated=factor(as.character(Hypomethylated), levels = group_name))
  mdat_sig <- mdat %>% filter(Hypomethylated != "Nonsignif") #%>%
    #dplyr::slice(1:100)
  
  # beta value %methylation
  edata <- data.frame(exprs(profile))
  mdat_bVals <- inner_join(mdat_sig , 
                           edata %>% rownames_to_column("GeneID"),
                           by = "GeneID")
  
  plot_volinFun <- function(tag){
    
    # tag=group_name[1]
    
    dat_tag <- mdat_bVals %>% filter(Hypomethylated==tag) %>%
      dplyr::select(UCSC_RefGene_Group, starts_with("P0")) %>%
      mutate(UCSC_RefGene_Group=factor(as.character(UCSC_RefGene_Group), 
                  levels = c("TSS1500","TSS200","5\'UTR","1stExon", "Body", "ExonBnd", "3\'UTR"))) 
    # summary the number of genic regions
    dat_status <- table(dat_tag$UCSC_RefGene_Group)
    print(dat_status)
    
    xaxis_lab <- c()
    for(i in 1:length(dat_status)){
      xaxis_lab <- c(xaxis_lab, paste0(names(dat_status)[i], "\n(", as.numeric(dat_status)[i], ")"))
    }
    
    predata <-  dat_tag %>%
      tidyr::gather(key="SampleID", value="bVals", -UCSC_RefGene_Group)
    
    pheno <- pData(profile) %>% dplyr::select(Type) %>%
      rownames_to_column("SampleID")
    
    plotdata <- inner_join(pheno, predata, by="SampleID") %>%
      mutate(Type=factor(as.character(Type), levels = group_name))
    
    # Grouped violin plot with split violins
    GeomSplitViolin <- ggproto("GeomSplitViolin", GeomViolin, 
                               draw_group = function(self, data, ..., draw_quantiles = NULL) {
      data <- transform(data, xminv = x - violinwidth * (x - xmin), xmaxv = x + violinwidth * (xmax - x))
      grp <- data[1, "group"]
      newdata <- plyr::arrange(transform(data, x = if (grp %% 2 == 1) xminv else xmaxv), if (grp %% 2 == 1) y else -y)
      newdata <- rbind(newdata[1, ], newdata, newdata[nrow(newdata), ], newdata[1, ])
      newdata[c(1, nrow(newdata) - 1, nrow(newdata)), "x"] <- round(newdata[1, "x"])
    
      if (length(draw_quantiles) > 0 & !scales::zero_range(range(data$y))) {
        stopifnot(all(draw_quantiles >= 0), all(draw_quantiles <=
          1))
        quantiles <- ggplot2:::create_quantile_segment_frame(data, draw_quantiles)
        aesthetics <- data[rep(1, nrow(quantiles)), setdiff(names(data), c("x", "y")), drop = FALSE]
        aesthetics$alpha <- rep(1, nrow(quantiles))
        both <- cbind(quantiles, aesthetics)
        quantile_grob <- GeomPath$draw_panel(both, ...)
        ggplot2:::ggname("geom_split_violin", grid::grobTree(GeomPolygon$draw_panel(newdata, ...), quantile_grob))
      }
      else {
        ggplot2:::ggname("geom_split_violin", GeomPolygon$draw_panel(newdata, ...))
      }
    })  
    
    geom_split_violin <- function(mapping = NULL, 
                                  data = NULL, 
                                  stat = "ydensity", 
                                  position = "identity", 
                                  ..., 
                                  draw_quantiles = NULL, 
                                  trim = TRUE, 
                                  scale = "area", 
                                  na.rm = FALSE, 
                                  show.legend = NA, 
                                  inherit.aes = TRUE) {
      layer(data = data, 
            mapping = mapping, 
            stat = stat, 
            geom = GeomSplitViolin, 
            position = position, 
            show.legend = show.legend, 
            inherit.aes = inherit.aes, 
            params = list(trim = trim, 
                          scale = scale, 
                          draw_quantiles = draw_quantiles, 
                          na.rm = na.rm, 
                          ...))
    }
   
    pl <- ggplot(plotdata, aes(x=UCSC_RefGene_Group, y=bVals, fill = Type))+ 
            geom_split_violin()+
            ggtitle(paste0("Methylation level by genic region", " (", tag, " genes)"))+
            xlab("")+
            ylab("%Methylation")+
            scale_fill_manual(name = NULL,
                               values = group_col,
                               labels = group_name)+
            scale_y_continuous(labels = scales::percent, limits=c(0, 1.2))+
            scale_x_discrete(breaks=c("TSS1500","TSS200","5\'UTR","1stExon","Body","ExonBnd","3\'UTR"),
                             labels=xaxis_lab)+
            theme_minimal()+
            theme(plot.title = element_text(color = "black", size = 12, face = "bold", hjust = .5),
                  axis.title = element_text(color = "black", size = 10),
                  axis.text = element_text(color = "black", size = 10),
                  text = element_text(size = 8, color = "black", family="serif"),
                  #panel.grid = element_blank(),
                  legend.position = "bottom",
                  legend.key.height = unit(0.6, "cm"),
                  legend.text = element_text(face = "bold", color = "black", size = 8)) 
    return(pl)
  }

  p1 <- plot_volinFun(tag = group_name[1])
  p2 <- plot_volinFun(tag = group_name[2])
  
  res <- list(grp1=p1, grp2=p2)
}
}

# ParacancerousTissue versus PrimaryCancerTissue
DMP_violin_PPC <- DMP_violin(profile=prf_PPC,
           dataDMP=DMP_PPC,
           group_name=c("ParacancerousTissue", "PrimaryCancerTissue"),
           group_col=c("#397A60", "#FC6940"))

DMP_violin_PPC_v2 <- cowplot::plot_grid(DMP_violin_PPC$grp1, 
                                        DMP_violin_PPC$grp2, 
                   ncol = 1, align = "hv")
DMP_violin_PPC_v2
if(!dir.exists("../../Result/Figure")){
  dir.create("../../Result/Figure")
}
ggsave("../../Result/Figure/Paracancerous_PrimaryCancer_DMP_violin.pdf", DMP_violin_PPC_v2, width = 8, height = 8)


# PrimaryCancerTissue versus LiverMetastasis
DMP_violin_PCL <- DMP_violin(profile=prf_PCL,
           dataDMP=DMP_PCL,
           group_name=c("PrimaryCancerTissue", "LiverMetastasis"),
           group_col=c("#397A60", "#FC6940"))

DMP_violin_PCL_v2 <- cowplot::plot_grid(DMP_violin_PCL$grp1, DMP_violin_PCL$grp2, 
                   ncol = 1, align = "hv")
DMP_violin_PCL_v2
if(!dir.exists("../../Result/Figure")){
  dir.create("../../Result/Figure")
}
ggsave("../../Result/Figure/PrimaryCancer_Liver_DMP_violin.pdf", DMP_violin_PCL_v2, width = 8, height = 8)
```


### The overlap between upregulated gene and hypomethylated loci at promoters and enhancer
```{r, fig.width=9, fig.height=7}
# get profile and phenotype
get_overlap_DMP_DEG <- function(
                        DMP_Meth=DMP_PPC,
                        DEG_RNA=DEG_PPC,
                        RefGene_Group=c("TSS1500","TSS200"),
                        group_name=c("ParacancerousTissue", "PrimaryCancerTissue")){
  
  # DMP_Meth=DMP_PPC
  # DEG_RNA=DEG_PPC
  # RefGene_Group=c("TSS1500", "TSS200")
  # group_name=c("ParacancerousTissue", "PrimaryCancerTissue")
  
  # maintain unique name of UCSC_RefGene_Name and UCSC_RefGene_Group
  require(DEP)
  gene_uniq_name <- make_unique(DMP_Meth, "UCSC_RefGene_Name", "UCSC_RefGene_Group", delim = ";") %>%
    dplyr::select(GeneID, name, ID) %>%
    filter(!name %in% grep("^.\\d", name, value=T)) %>%
    filter(name != "", ID != "") %>%
    setNames(c("GeneID", "UCSC_RefGene_Name", "UCSC_RefGene_Group")) %>%
    filter(UCSC_RefGene_Group %in% RefGene_Group) %>%
    mutate(Gene=gsub("\\.\\d+$", "", UCSC_RefGene_Name),
           RefGene=paste0(UCSC_RefGene_Group, "_", Gene))
  
  # merge data 
  mdat <- inner_join(gene_uniq_name, 
                     DMP_Meth %>% 
                       dplyr::select(GeneID, Block, logFC, adj.P.Val,
                                     Hypomethylated, AveExpr, t, P.Value, B, Delta),
                     by ="GeneID")
  mdat_sig <- mdat %>% filter(Hypomethylated %in% group_name) 
  
  # deduplicate methylated genes & intersection between RNA-seq and DNA methylation
  sig_dedup_methy <- mdat_sig[duplicated(mdat_sig$Gene), ]
  
  # overlap genes: Upregulated genes vs Hypomethylated DMRs at promoters in ParacancerousTissue
  sig_gene_grp1 <- DEG_RNA %>% filter(Enrichment == group_name[1])
  sig_dedup_methy_grp2 <- sig_dedup_methy %>% filter(Hypomethylated %in% group_name[2])
  genegrp1_methygrp2 <- intersect(sig_dedup_methy_grp2$Gene, sig_gene_grp1$GeneID)
  overlap_methy_grp2 <- sig_dedup_methy_grp2 %>% filter(Gene%in%genegrp1_methygrp2)  
  overlap_gene_grp1 <- DEG_RNA %>% filter(GeneID %in% genegrp1_methygrp2) %>%
    arrange(desc(abs(logFC)))
  print(overlap_gene_grp1$GeneID)
  
  library(ggvenn)
  plotlist <- list("Upregulated genes"=sig_gene_grp1$GeneID,
                   "Hypomethylated \nDMRs at promoters"=sig_dedup_methy_grp2$Gene)
  
  para_venn <- ggvenn(plotlist,
                columns = c("Upregulated genes", "Hypomethylated \nDMRs at promoters"),
                stroke_size = 0.5, 
                set_name_size = 4)+
    ggtitle(group_name[1])+
    theme(plot.title = element_text(hjust = .5, size = 10, face = "bold"),
          text = element_text(size = 8, color = "black", family="serif"))

  # overlap genes: Upregulated genes vs Hypomethylated DMRs at promoters in PrimaryCancerTissue  
  sig_gene_grp2 <- DEG_RNA %>% filter(Enrichment == group_name[2])
  sig_dedup_methy_grp1 <- sig_dedup_methy %>% filter(Hypomethylated %in% group_name[1])    
  genegrp2_methygrp1 <- intersect(sig_dedup_methy_grp1$Gene, sig_gene_grp2$GeneID)  
  overlap_methy_grp1 <- sig_dedup_methy_grp2 %>% filter(Gene%in%genegrp2_methygrp1)  
  overlap_gene_grp2 <- DEG_RNA %>% filter(GeneID %in% genegrp2_methygrp1) %>%
    arrange(desc(abs(logFC)))
  print(overlap_gene_grp2$GeneID)
  
  plotlist2 <- list("Upregulated genes"=sig_gene_grp2$GeneID,
                   "Hypomethylated \nDMRs at promoters"=sig_dedup_methy_grp1$Gene)
  
  prim_venn <- ggvenn(plotlist2,
                columns = c("Upregulated genes", "Hypomethylated \nDMRs at promoters"),
                stroke_size = 0.5, 
                set_name_size = 4)+
    ggtitle(group_name[2])+
    theme(plot.title = element_text(hjust = .5, size = 10, face = "bold"),
          text = element_text(size = 8, color = "black", family="serif"))
  
  res <- cowplot::plot_grid(para_venn, 
                            prim_venn, 
                            nrow = 1,
                            align = "hv")
  
  return(res)
}

# methylated Gene boxplot or barplot
paired_boxplot <- function(
        profile=prf_PPC,
        dataDMP=DMP_PPC,
        genelist=c("UGT1A10","KRT16","SERPINB5","TNS4","TFF1","KRT6A","CAPN9","CDHR2","HTR1D","MMP13",
                    "CLPS","CELA2A","GDF10","ERP27","LHCGR","CTRB2","LEP","KLK1","SERPINI2","PEX5L"),
        group_name=c("ParacancerousTissue", "PrimaryCancerTissue"),
        RefGene_Group=c("TSS1500", "TSS200"),
        group_col=c("#FF7F50", "#7CCD7C")){

  # profile=prf_PPC
  # dataDMP=DMP_PPC
  # genelist=c("UGT1A10","KRT16","SERPINB5","TNS4","TFF1","KRT6A","CAPN9","CDHR2","HTR1D","MMP13",
  #            "CLPS","CELA2A","GDF10","ERP27","LHCGR","CTRB2","LEP","KLK1","SERPINI2","PEX5L")
  # group_name=c("ParacancerousTissue", "PrimaryCancerTissue")
  # RefGene_Group=c("TSS1500", "TSS200")
  # group_col=c("#FF7F50", "#7CCD7C")
  
  # dissect the methylation patterns of differentially expressed genes
  require(DEP)
  gene_uniq_name <- make_unique(dataDMP, "UCSC_RefGene_Name", "UCSC_RefGene_Group", delim = ";") %>%
    dplyr::select(GeneID, name, ID) %>%
    filter(!name %in% grep("^.\\d", name, value=T)) %>%
    filter(name != "", ID != "") %>%
    setNames(c("GeneID", "UCSC_RefGene_Name", "UCSC_RefGene_Group")) %>%
    filter(UCSC_RefGene_Group %in% RefGene_Group) %>%
    mutate(Gene=gsub("\\.\\d+$", "", UCSC_RefGene_Name),
           RefGene=paste0(UCSC_RefGene_Group, "_", Gene))
  
  # merge data 
  mdat <- inner_join(gene_uniq_name, 
                     dataDMP %>% 
                       dplyr::select(GeneID, Block, logFC, adj.P.Val,
                                     Hypomethylated, AveExpr, t, P.Value, B, Delta),
                     by ="GeneID")
  cpg_gene <- mdat %>% filter(Hypomethylated %in% group_name) %>%
    filter(Gene%in%genelist) %>%
    dplyr::select(GeneID,  Gene)
  
  # group information
  pheno <- pData(profile)
  pheno$Type <- factor(as.character(pheno$Type), levels = group_name)
  pheno$PID <- factor(as.character(pheno$PID)) 
  
  # profile
  edata <- data.frame(exprs(profile)) %>%
    rownames_to_column("GeneID") 
  prof <- inner_join(cpg_gene, edata, by="GeneID") %>%
    dplyr::select(-GeneID) %>%
    group_by(Gene) %>%
    summarise_each(median) %>%
    column_to_rownames("Gene")
  
  mdat <- pheno %>% dplyr::select(PID, Type) %>%
    rownames_to_column("SampleID") %>%
    inner_join(t(prof) %>% data.frame() %>% rownames_to_column("SampleID"), by = "SampleID") %>%
    column_to_rownames("SampleID")
  
  plotdata <- mdat %>% tidyr::gather(key="geneID", value="value", -c("PID", "Type")) 
  
  # arrange by median of group per genes 
  # plotdata_median_delta <- plotdata %>% group_by(Type, geneID) %>%
  #   summarise(median_value=median(value)) %>%
  #   ungroup() %>%
  #   arrange(geneID, Type, median_value) %>%
  #   group_by(geneID) %>%
  #   summarise(delta=dplyr::first(median_value) - dplyr::last(median_value)) %>%
  #   arrange(delta)
  # 
  # print(unique(plotdata$PID))
  
  plotdata$geneID <- factor(plotdata$geneID, levels = genelist)
  
  pl <- ggplot(plotdata, aes(x = Type, y = value, fill= Type))+
    stat_boxplot(geom = "errorbar", width = 0.15,
                 position = position_dodge(0.4)) +    
    geom_boxplot(width = 0.4, 
                 outlier.colour = "black", 
                 outlier.shape=21, outlier.size = 1)+
    geom_line(aes(group=PID), linetype=2, size=.5)+
    scale_fill_manual(values = group_col)+
    scale_y_continuous(labels = scales::percent, limits=c(0, 1))+    
    #facet_wrap(facets = "geneID", scales = "free_x", nrow = 2)+
    facet_wrap(facets = "geneID", nrow = 2)+    
    labs(x="", y="%Methylation")+
    guides(fill=F)+
    theme_classic()+
    theme(axis.title = element_text(color = "black", size = 12),
          axis.text.x = element_text(color = "black", size = 10, angle = 30, hjust = .5, vjust = .5),
          text = element_text(size = 8, color = "black", family="serif"),
          panel.grid = element_blank(),
          strip.text = element_text(face = "bold", size = 12))    
    
  return(pl)
}

# ParacancerousTissue versus PrimaryCancerTissue 
# ParacancerousTissue
PPC_venn <- get_overlap_DMP_DEG(
                    DMP_Meth=DMP_PPC,
                    DEG_RNA=DEG_PPC,
                    RefGene_Group=c("TSS1500","TSS200"),
                    group_name=c("ParacancerousTissue", "PrimaryCancerTissue"))
PPC_venn
if(!dir.exists("../../Result/Figure")){
  dir.create("../../Result/Figure")
}
ggsave("../../Result/Figure/Paracancerous_PrimaryCancer_methylatedgene_venn.pdf", PPC_venn, width = 8, height = 8)


boxplot_PPC <- paired_boxplot(
        profile=prf_PPC,
        dataDMP=DMP_PPC,
        genelist=c("UGT1A10","KRT16","SERPINB5","TNS4","TFF1","KRT6A","CAPN9","CDHR2","HTR1D","MMP13",
                   "CLPS","CELA2A","GDF10","ERP27","LHCGR","CTRB2","LEP","KLK1","SERPINI2","PEX5L"),
        group_name=c("ParacancerousTissue", "PrimaryCancerTissue"))
boxplot_PPC
if(!dir.exists("../../Result/Figure")){
  dir.create("../../Result/Figure")
}
ggsave("../../Result/Figure/Paracancerous_PrimaryCancer_MethylatedGene.pdf", boxplot_PPC, width = 8, height = 8)


# PrimaryCancerTissue versus LiverMetastasis
# ParacancerousTissue
PCL_venn <- get_overlap_DMP_DEG(
                    DMP_Meth=DMP_PCL,
                    DEG_RNA=DEG_PCL,
                    RefGene_Group=c("TSS1500","TSS200"),
                    group_name=c("PrimaryCancerTissue", "LiverMetastasis"))
PCL_venn
if(!dir.exists("../../Result/Figure")){
  dir.create("../../Result/Figure")
}
ggsave("../../Result/Figure/PrimaryCancer_Liver_methylatedgene_venn.pdf", PCL_venn, width = 8, height = 8)


boxplot_PCL <- paired_boxplot(
        profile=prf_PCL,
        dataDMP=DMP_PCL,
        genelist=c("MASP2","FGA","KNG1","FABP1","CFHR5","HAO1","SERPIND1","SLC10A1","CYP8B1","ARG1",
                   "DLX5","LRRN4CL","SPARCL1","DPT","TRIL","GLT8D2","CYS1","CCND2","CD34","TCF4"),
        group_name=c("PrimaryCancerTissue", "LiverMetastasis"))
boxplot_PCL
if(!dir.exists("../../Result/Figure")){
  dir.create("../../Result/Figure")
}
ggsave("../../Result/Figure/PrimaryCancer_Liver_MethylatedGene.pdf", boxplot_PCL, width = 8, height = 8)
```


### DMRs 5/17/2021

* Hypomethylated CpG sites or Regions

* Hypermethylated CpG sites or Regions
```{r}
# methylation DMRs
DMR_PPC <- readRDS("../../Result/Differential/Paracancerous_PrimaryCancer_DMR_bVals.RDS")
DMR_PCL <- readRDS("../../Result/Differential/PrimaryCancer_Liver_DMR_bVals.RDS")


TopDMR_plot <- function(dat=DMR_PPC, topN=1){
  
  # dat <- DMR_PPC
  # topN <- 2
  
  targets <- pData(dat$Datset)
  results.ranges <- dat$Range
  bVals <- exprs(dat$Datset)
  
  # set up the grouping variables and colours  
  pal <- RColorBrewer::brewer.pal(8,"Dark2")
  groups <- pal[1:length(unique(targets$Type))]
  names(groups) <- levels(factor(targets$Type))
  cols <- groups[as.character(factor(targets$Type))]
  
  # draw the plot for the top DMR
  par(mfrow=c(1, 1))
  DMR.plot(ranges = results.ranges, 
           dmr = topN, 
           CpGs = bVals, 
           phen.col = cols, 
           what = "Beta", 
           arraytype = "EPIC", 
           genome = "hg19")
}

# Top1 DMR in Paracancerous_PrimaryCancer
TopDMR_plot(dat=DMR_PPC, topN=1)

# Top1 DMR in PrimaryCancer_Liver
TopDMR_plot(dat=DMR_PCL, topN=1)
```


### Hypomethylated DMRs at promoters and Hypomethylated DMRs at intronic enhancers (5/26/2021)

We met a problem that there were multiple gene targeted by one probe site. Therefore, how to dissect the DNA methylated array matrix would be very vital for downstream analysis. The first approach is to select the first gene name of DMRs in the ordered rank, and the other approach is to attain all the genes with the same probe site.

This function as the following shows two methods to dissect DNA methylated matrix
```{r}
DMR_PPC_hypoPromoter <- fread("../../Result/Hypomethylated/DMR_PPC_hypomethylatedPromoter.csv")
DMR_PPC_hypoEnhancer <- fread("../../Result/Hypomethylated/DMR_PPC_hypomethylatedEnhancer.csv")
DMR_PCL_hypoPromoter <- fread("../../Result/Hypomethylated/DMR_PCL_hypomethylatedPromoter.csv")
DMR_PCL_hypoEnhancer <- fread("../../Result/Hypomethylated/DMR_PCL_hypomethylatedEnhancer.csv")


get_DMP_DEG <- function(DMR=DMR_PPC_hypoPromoter,
                        DEG_RNA=DEG_PPC,
                        group_name=c("ParacancerousTissue", "PrimaryCancerTissue"),
                        subGroup="promoters",
                        Method="single"){
  
  # DMR=DMR_PPC_hypoPromoter
  # DEG_RNA=DEG_PPC
  # group_name=c("ParacancerousTissue", "PrimaryCancerTissue")
  # subGroup="promoters"
  # Method="single"
  
  # maintain unique name 
  methy_grp1 <- DMR %>% filter(Hypomethylated %in% group_name[1])
  methy_grp2 <- DMR %>% filter(Hypomethylated %in% group_name[2])
  if(Method == "single"){
    methy_grp1_gene <- unique(as.character(sapply(as.vector(methy_grp1$MatchGene), 
                                           function(x){unlist(strsplit(x, ","))[1]})))
    methy_grp2_gene <- unique(as.character(sapply(as.vector(methy_grp2$MatchGene), 
                                           function(x){unlist(strsplit(x, ","))[1]})))     
  }else if(Method == "multiple"){
    methy_grp1_gene <- unique(unlist(strsplit(methy_grp1$MatchGene, ",")))
    methy_grp2_gene <- unique(unlist(strsplit(methy_grp2$MatchGene, ",")))     
  }
  
  # overlap genes: Upregulated genes vs Hypomethylated DMRs 
  gene_grp1 <- DEG_RNA %>% filter(Enrichment == group_name[1])
  gene1_methy1 <- intersect(gene_grp1$GeneID, methy_grp1_gene)
  gene_grp2 <- DEG_RNA %>% filter(Enrichment == group_name[2])
  gene2_methy2 <- intersect(gene_grp2$GeneID, methy_grp2_gene)  
  
  library(ggvenn)
  if(subGroup == "promoters"){
    plotlist <- list("Upregulated genes"=gene_grp1$GeneID,
                     "Hypomethylated \nDMRs at promoters"=methy_grp1_gene)
    
    venn1 <- ggvenn(plotlist,
                  columns = c("Upregulated genes", "Hypomethylated \nDMRs at promoters"),
                  stroke_size = 0.5, 
                  set_name_size = 4)+
      ggtitle(group_name[1])+
      theme(plot.title = element_text(hjust = .5, size = 10, face = "bold"),
            text = element_text(size = 8, color = "black", family="serif")) 
    plotlist2 <- list("Upregulated genes"=gene_grp2$GeneID,
                     "Hypomethylated \nDMRs at promoters"=methy_grp2_gene)
    
    venn2 <- ggvenn(plotlist2,
                  columns = c("Upregulated genes", "Hypomethylated \nDMRs at promoters"),
                  stroke_size = 0.5, 
                  set_name_size = 4)+
      ggtitle(group_name[2])+
      theme(plot.title = element_text(hjust = .5, size = 10, face = "bold"),
            text = element_text(size = 8, color = "black", family="serif"))    
  }else{
    plotlist <- list("Upregulated genes"=gene_grp1$GeneID,
                     "Hypomethylated \nDMRs at intronic enhancers"=methy_grp1_gene)
    
    venn1 <- ggvenn(plotlist,
                  columns = c("Upregulated genes", "Hypomethylated \nDMRs at intronic enhancers"),
                  stroke_size = 0.5, 
                  set_name_size = 4)+
      ggtitle(group_name[1])+
      theme(plot.title = element_text(hjust = .5, size = 10, face = "bold"),
            text = element_text(size = 8, color = "black", family="serif")) 
    plotlist2 <- list("Upregulated genes"=gene_grp2$GeneID,
                     "Hypomethylated \nDMRs at intronic enhancers"=methy_grp2_gene)
    
    venn2 <- ggvenn(plotlist2,
                  columns = c("Upregulated genes", "Hypomethylated \nDMRs at intronic enhancers"),
                  stroke_size = 0.5, 
                  set_name_size = 4)+
      ggtitle(group_name[2])+
      theme(plot.title = element_text(hjust = .5, size = 10, face = "bold"),
            text = element_text(size = 8, color = "black", family="serif"))     
  }
  
  # venn plot
  res_venn <- cowplot::plot_grid(venn1, 
                            venn2, 
                            nrow = 1,
                            align = "hv")
  
  # overlap genes
  gene_overlap <- c(gene1_methy1, gene2_methy2)
  
  
  # DMRs 
  res_DMRs <- DMR  
  res_DMRs$GeneOverlap <- NA
  for(i in 1:nrow(res_DMRs)){
    gene_char <- unlist(strsplit(res_DMRs$MatchGene[i], ","))
    if(length(gene_char) > 1){
      for(j in 1:length(gene_char)){
        index <- grep(paste0("^", gene_char[j], "$"), gene_overlap, perl = T)
      }
    }else{
      index <- grep(paste0("^", gene_char, "$"), gene_overlap, perl = T)
    }
    if(length(index) > 0){
      res_DMRs$GeneOverlap[i] <- TRUE
    }else{
      res_DMRs$GeneOverlap[i] <- FALSE
    }
  }
  res_DMRs <- res_DMRs[res_DMRs$GeneOverlap != FALSE, ] %>%
    dplyr::select(-GeneOverlap)  
  
  # DEGs
  res_DEGs <- DEG_RNA
  res_DEGs$GeneOverlap <- NA
  for(i in 1:nrow(res_DEGs)){
    index <- grep(paste0("^", res_DEGs$GeneID[i], "$"), gene_overlap, perl = T)
    if(length(index) > 0){
      res_DEGs$GeneOverlap[i] <- TRUE
    }else{
      res_DEGs$GeneOverlap[i] <- FALSE
    }
  }  
  res_DEGs <- res_DEGs[res_DEGs$GeneOverlap != FALSE, ] %>%
    dplyr::select(-GeneOverlap)
  
  res <- list(venn=res_venn,
              DMRs=res_DMRs,
              DEGs=res_DEGs)
  
  return(res)
}

PPC_Promoter_single <- get_DMP_DEG(
  DMR=DMR_PPC_hypoPromoter,
  DEG_RNA=DEG_PPC,
  group_name=c("ParacancerousTissue", "PrimaryCancerTissue"),
  subGroup="promoters",
  Method="single")
PPC_Promoter_single$venn

PPC_Enhancer_single <- get_DMP_DEG(
  DMR=DMR_PPC_hypoEnhancer,
  DEG_RNA=DEG_PPC,
  group_name=c("ParacancerousTissue", "PrimaryCancerTissue"),
  subGroup="enhancers",
  Method="single")
PPC_Enhancer_single$venn


PCL_Promoter_single <- get_DMP_DEG(
  DMR=DMR_PCL_hypoPromoter,
  DEG_RNA=DEG_PCL,
  group_name=c("PrimaryCancerTissue", "LiverMetastasis"),
  subGroup="promoters",
  Method="single")
PCL_Promoter_single$venn

PCL_Enhancer_single <- get_DMP_DEG(
  DMR=DMR_PCL_hypoEnhancer,
  DEG_RNA=DEG_PCL,
  group_name=c("PrimaryCancerTissue", "LiverMetastasis"),
  subGroup="enhancers",
  Method="single")
PCL_Promoter_single$venn

if(!dir.exists("../../Result/Figure")){
  dir.create("../../Result/Figure")
}
ggsave("../../Result/Figure/Paracancerous_PrimaryCancer_hypoPromoter_single_venn.pdf", 
       PPC_Promoter_single$venn, width = 8, height = 8)
ggsave("../../Result/Figure/Paracancerous_PrimaryCancer_hypoEnhancer_single_venn.pdf", 
       PPC_Enhancer_single$venn, width = 8, height = 8)
ggsave("../../Result/Figure/PrimaryCancer_Liver_hypoPromoter_single_venn.pdf", 
       PCL_Promoter_single$venn, width = 8, height = 8)
ggsave("../../Result/Figure/PrimaryCancer_Liver_hypoEnhancer_single_venn.pdf", 
       PCL_Enhancer_single$venn, width = 8, height = 8)

if(!dir.exists("../../Result/Overlap_gene/")){
  dir.create("../../Result/Overlap_gene/")
}

write.csv(PPC_Promoter_single$DMRs, "../../Result/Overlap_gene/Paracancerous_PrimaryCancer_hypoPromoter_single_DMRs.csv",
          row.names = F)
write.csv(PPC_Enhancer_single$DMRs, "../../Result/Overlap_gene/Paracancerous_PrimaryCancer_hypoEnhancer_single_DMRs.csv",
          row.names = F)
write.csv(PCL_Promoter_single$DMRs, "../../Result/Overlap_gene/PrimaryCancer_Liver_hypoPromoter_single_DMRs.csv",
          row.names = F)
write.csv(PCL_Enhancer_single$DMRs, "../../Result/Overlap_gene/PrimaryCancer_Liver_hypoEnhancer_single_DMRs.csv",
          row.names = F)

write.csv(PPC_Promoter_single$DEGs, "../../Result/Overlap_gene/Paracancerous_PrimaryCancer_hypoPromoter_single_DEGs.csv",
          row.names = F)
write.csv(PPC_Enhancer_single$DEGs, "../../Result/Overlap_gene/Paracancerous_PrimaryCancer_hypoEnhancer_single_DEGs.csv",
          row.names = F)
write.csv(PCL_Promoter_single$DEGs, "../../Result/Overlap_gene/PrimaryCancer_Liver_hypoPromoter_single_DEGs.csv",
          row.names = F)
write.csv(PCL_Enhancer_single$DEGs, "../../Result/Overlap_gene/PrimaryCancer_Liver_hypoEnhancer_single_DEGs.csv",
          row.names = F)



PPC_Promoter_multiple <- get_DMP_DEG(
  DMR=DMR_PPC_hypoPromoter,
  DEG_RNA=DEG_PPC,
  group_name=c("ParacancerousTissue", "PrimaryCancerTissue"),
  subGroup="promoters",
  Method="multiple")


PPC_Enhancer_multiple <- get_DMP_DEG(
  DMR=DMR_PPC_hypoEnhancer,
  DEG_RNA=DEG_PPC,
  group_name=c("ParacancerousTissue", "PrimaryCancerTissue"),
  subGroup="enhancers",
  Method="multiple")


PCL_Promoter_multiple <- get_DMP_DEG(
  DMR=DMR_PCL_hypoPromoter,
  DEG_RNA=DEG_PCL,
  group_name=c("PrimaryCancerTissue", "LiverMetastasis"),
  subGroup="promoters",
  Method="multiple")


PCL_Enhancer_multiple <- get_DMP_DEG(
  DMR=DMR_PCL_hypoEnhancer,
  DEG_RNA=DEG_PCL,
  group_name=c("PrimaryCancerTissue", "LiverMetastasis"),
  subGroup="enhancers",
  Method="multiple")


if(!dir.exists("../../Result/Figure")){
  dir.create("../../Result/Figure")
}
ggsave("../../Result/Figure/Paracancerous_PrimaryCancer_hypoPromoter_multiple_venn.pdf", 
       PPC_Promoter_multiple$venn, width = 8, height = 8)
ggsave("../../Result/Figure/Paracancerous_PrimaryCancer_hypoEnhancer_multiple_venn.pdf", 
       PPC_Enhancer_multiple$venn, width = 8, height = 8)
ggsave("../../Result/Figure/PrimaryCancer_Liver_hypoPromoter_multiple_venn.pdf", 
       PCL_Promoter_multiple$venn, width = 8, height = 8)
ggsave("../../Result/Figure/PrimaryCancer_Liver_hypoEnhancer_multiple_venn.pdf", 
       PCL_Enhancer_multiple$venn, width = 8, height = 8)

if(!dir.exists("../../Result/Overlap_gene/")){
  dir.create("../../Result/Overlap_gene/")
}

write.csv(PPC_Promoter_multiple$DMRs,
          "../../Result/Overlap_gene/Paracancerous_PrimaryCancer_hypoPromoter_multiple_DMRs.csv",
          row.names = F)
write.csv(PPC_Enhancer_multiple$DMRs,
          "../../Result/Overlap_gene/Paracancerous_PrimaryCancer_hypoEnhancer_multiple_DMRs.csv",
          row.names = F)
write.csv(PCL_Promoter_multiple$DMRs, 
          "../../Result/Overlap_gene/PrimaryCancer_Liver_hypoPromoter_multiple_DMRs.csv",
          row.names = F)
write.csv(PCL_Enhancer_multiple$DMRs, 
          "../../Result/Overlap_gene/PrimaryCancer_Liver_hypoEnhancer_multiple_DMRs.csv",
          row.names = F)

write.csv(PPC_Promoter_multiple$DEGs, 
          "../../Result/Overlap_gene/Paracancerous_PrimaryCancer_hypoPromoter_multiple_DEGs.csv",
          row.names = F)
write.csv(PPC_Enhancer_multiple$DEGs, 
          "../../Result/Overlap_gene/Paracancerous_PrimaryCancer_hypoEnhancer_multiple_DEGs.csv",
          row.names = F)
write.csv(PCL_Promoter_multiple$DEGs, 
          "../../Result/Overlap_gene/PrimaryCancer_Liver_hypoPromoter_multiple_DEGs.csv",
          row.names = F)
write.csv(PCL_Enhancer_multiple$DEGs, 
          "../../Result/Overlap_gene/PrimaryCancer_Liver_hypoEnhancer_multiple_DEGs.csv",
          row.names = F)
```


```{r}
# methylated Gene boxplot or barplot
paired_boxplot_v2 <- function(
        profile=prf_PPC,
        DEGs=PPC_Enhancer_multiple$DEGs,
        genelist=c("CDH3","ONECUT3","TRIM15"),
        group_name=c("ParacancerousTissue", "PrimaryCancerTissue"),
        group_col=c("#FF7F50", "#7CCD7C")){

  profile=prf_PPC
  DEG=PPC_Enhancer_multiple$DEGs
  genelist=c("CDH3","ONECUT3","TRIM15")
  group_name=c("ParacancerousTissue", "PrimaryCancerTissue")
  RefGene_Group=c("TSS1500", "TSS200")
  group_col=c("#FF7F50", "#7CCD7C")

  sid <- intersect(genelist, DEG$GeneID)
  mdat <- reannEPIC[reannEPIC$Gene_Name%in%sid, c("Name", "Gene_Name")] %>%
    dplyr::rename(GeneID=Name) %>%
    dplyr::rename(Gene=Gene_Name)
  
  # group information
  pheno <- pData(profile)
  pheno$Type <- factor(as.character(pheno$Type), levels = group_name)
  pheno$PID <- factor(as.character(pheno$PID)) 
  
  # profile
  edata <- data.frame(exprs(profile)) %>%
    rownames_to_column("GeneID") 
  prof <- inner_join(mdat, edata, by="GeneID") %>%
    dplyr::select(-GeneID) %>%
    group_by(Gene) %>%
    summarise_each(median) %>%
    column_to_rownames("Gene")
  
  mdat <- pheno %>% dplyr::select(PID, Type) %>%
    rownames_to_column("SampleID") %>%
    inner_join(t(prof) %>% data.frame() %>% rownames_to_column("SampleID"), by = "SampleID") %>%
    column_to_rownames("SampleID")
  
  plotdata <- mdat %>% tidyr::gather(key="geneID", value="value", -c("PID", "Type")) 
  
  # arrange by median of group per genes 
  # plotdata_median_delta <- plotdata %>% group_by(Type, geneID) %>%
  #   summarise(median_value=median(value)) %>%
  #   ungroup() %>%
  #   arrange(geneID, Type, median_value) %>%
  #   group_by(geneID) %>%
  #   summarise(delta=dplyr::first(median_value) - dplyr::last(median_value)) %>%
  #   arrange(delta)
  # 
  # print(unique(plotdata$PID))
  
  plotdata$geneID <- factor(plotdata$geneID, levels = genelist)
  
  pl <- ggplot(plotdata, aes(x = Type, y = value, fill= Type))+
    stat_boxplot(geom = "errorbar", width = 0.15,
                 position = position_dodge(0.4)) +    
    geom_boxplot(width = 0.4, 
                 outlier.colour = "black", 
                 outlier.shape=21, outlier.size = 1)+
    geom_line(aes(group=PID), linetype=2, size=.5)+
    scale_fill_manual(values = group_col)+
    scale_y_continuous(labels = scales::percent, limits=c(0, 1))+    
    #facet_wrap(facets = "geneID", scales = "free_x", nrow = 2)+
    facet_wrap(facets = "geneID", nrow = 2)+    
    labs(x="", y="%Methylation")+
    guides(fill=F)+
    theme_classic()+
    theme(axis.title = element_text(color = "black", size = 12),
          axis.text.x = element_text(color = "black", size = 10, angle = 30, hjust = .5, vjust = .5),
          text = element_text(size = 8, color = "black", family="serif"),
          panel.grid = element_blank(),
          strip.text = element_text(face = "bold", size = 12))    
    
  return(pl)
}

paired_boxplot(
        profile=prf_PPC,
        dataDMP=DMP_PPC,
        genelist=c("CDH3"),
        group_name=c("ParacancerousTissue", "PrimaryCancerTissue"))
```


### systemic information
```{r}
sessionInfo()
```


### Reference

1. [Infinium MethylationEPIC Manifest Column Headings](https://support.illumina.com/content/dam/illumina-support/documents/downloads/productfiles/methylationepic/infinium-methylationepic-manifest-column-headings.pdf)

2. [Gene-Structure](http://chenyuan.date/2018/01/Gene-Structure/)

3. [Epigenetic regulation of the lineage specifcity of primary human dermal lymphatic and blood vascular endothelial cells](https://link.springer.com/content/pdf/10.1007/s10456-020-09743-9.pdf)

4. [Grouped violin plot with split violins](https://stackoverflow.com/questions/35717353/split-violin-plot-with-ggplot2)

