---
title: "Differential Expression Analysis: mRNA-Gene"
date: "Created: 2021-04-26 Updated: `r Sys.Date()`"
output: 
  html_notebook:
    codes: hide
---

### Introdution

Bulk-RNA data has been sequenced only in the three samples' resources, and some participants lacked of one or two, so I would do the EDA in two conditions:

1. **ParacancerousTissue** versus **PrimaryCancerTissue** (n=20, the paired participant);

2. **PrimaryCancerTissue** versus **LiverMetastasis** (n=12, the paired participant).

<details>
<summary>Figures of Data Distribution</summary>

<p align="center">
<img src="./images/Data_distribution.jpg" width="800" height="500">
</p>

<h4 align = "center"> Data Distribution in the population </h4>

</details>

```{r setup, include=FALSE}
knitr::opts_chunk$set(warning = F, message = F)
library(dplyr)
library(tibble)
library(ggplot2)
library(convert)
library(limma)
library(DESeq2)

# rm(list = ls())
options(stringsAsFactors = F)
options(future.globals.maxSize = 1000 * 1024^2)
```


### load data 
```{r}
genExprSet <- readRDS("../../Result/profile/GeneExprSet_filtered.RDS")
```


### DEA function

<details>
<summary>Function</summary>

1. get_paired_data

2. get_DiffGene_DESeq2

3. get_DiffGene_limma

</details>

```{r}
# get profile and phenotype
get_paired_data <- function(ExprSet   = genExprSet,
                            group_col = c("ParacancerousTissue", "PrimaryCancerTissue")){
  
  # ExprSet=genExprSet
  # group_col=c("ParacancerousTissue", "PrimaryCancerTissue") 
  
  phenotype <- pData(ExprSet)
  profile <- data.frame(exprs(ExprSet))
  
  phen_group1 <- phenotype %>% filter(Type%in%group_col[1])
  phen_group2 <- phenotype %>% filter(Type%in%group_col[2])
  pid <- intersect(phen_group1$PID, phen_group2$PID)
  phen <- phenotype %>% filter(Type%in%group_col) %>%
    filter(PID %in% pid)
  
  prof <- profile %>% dplyr::select(rownames(phen))
  
  # determine the right order between profile and phenotype 
  for(i in 1:ncol(prof)){ 
    if (!(colnames(prof)[i] == rownames(phen)[i])) {
      stop(paste0(i, " Wrong"))
    }
  }
  
  require(convert)
  exprs <- as.matrix(prof)
  adf <-  new("AnnotatedDataFrame", data=phen)
  experimentData <- new("MIAME",
        name="ShuiLin Liao", lab="Dong gdl Lab",
        contact="dong_ming@grmh-gdl.cn",
        title="Pancreatic tumor Experiment",
        abstract="The gene ExpressionSet",
        url="www.grmh-gdl.cn",
        other=list(notes="Chosen by groups"))
  expressionSet <- new("ExpressionSet", exprs=exprs,
                       phenoData=adf, 
                       experimentData=experimentData)
  
  return(expressionSet)
}

# Differential Expression Analysis by DESeq2
get_DiffGene_DESeq2 <- function(dataset=ExprSet1,
                                group_col=c("ParacancerousTissue", "PrimaryCancerTissue"),
                                pair=TRUE,
                                fc=1,
                                Pval=0.05){
  
  # dataset=ExprSet1
  # group_col=c("ParacancerousTissue", "PrimaryCancerTissue")
  # pair=TRUE
  # fc=1
  # Pval=0.05
  
  pheno <- pData(dataset)
  pheno$Type <- factor(as.character(pheno$Type), levels = group_col)
  pheno$PID <- factor(as.character(pheno$PID))
  edata <- as.matrix(exprs(dataset))
  if(pair){
    DESeqDataSet <- DESeqDataSetFromMatrix(
                        countData = edata, 
                        colData   = pheno,
                        design    =~ PID + Type)    
  }else{
    DESeqDataSet <- DESeqDataSetFromMatrix(
                        countData = edata, 
                        colData   = pheno,
                        design    =~ Type)    
  }

  dds <- DESeq(DESeqDataSet)
  print(dds)
  res <- results(dds, pAdjustMethod = "BH", alpha = 0.05) %>% na.omit()
  res <- res[order(res$padj), ] %>% 
    data.frame() %>%
    rownames_to_column("GeneID") %>%
    mutate(Block=resultsNames(dds)[length(resultsNames(dds))],
           Block=gsub("Type_", "", Block))
  
  # validate the enriched directory
  data.frame(edata)[rownames(data.frame(edata))%in%res$GeneID[1], , F] %>% 
    t() %>% data.frame() %>%
    setNames("Gene") %>%
    rownames_to_column("SampleID") %>%
    inner_join(pheno%>%rownames_to_column("SampleID"), by = "SampleID") -> a1
  wilcox.test(Gene~Type, data = a1, paired=T)
  ggplot(a1, aes(x=Type, y=Gene))+
    geom_boxplot()+
    labs(y=res$GeneID[1], x="")+
    geom_line(aes(group=PID))+
    theme_bw()
  
  # enriched directory
  res[which(res$log2FoldChange >= fc & res$padj < Pval), "Enrichment"] <- group_col[2]
  res[which(res$log2FoldChange <= -fc & res$padj < Pval), "Enrichment"] <- group_col[1]
  res[which(abs(res$log2FoldChange) < fc | res$padj >= Pval), "Enrichment"] <- "Nonsignif"
  res <- res[, c(1, 8:9, 2:7)]
  
  return(list(dds=dds, results=res))
}

# Differential Expression Analysis by limma
get_DiffGene_limma <- function(dataset=ExprSet1,
                               group_col=c("ParacancerousTissue", "PrimaryCancerTissue"),
                               pair=TRUE,
                               scale=TRUE,
                               fc=1,
                               Pval=0.05){

  # dataset=ExprSet1
  # group_col=c("ParacancerousTissue", "PrimaryCancerTissue")
  # pair=TRUE
  # scale=TRUE
  # fc=1
  # Pval=0.05
  
  pheno <- pData(dataset)
  pheno$Type <- factor(as.character(pheno$Type), levels = group_col)
  pheno$PID <- factor(as.character(pheno$PID))
  
  if(pair){
    # paired test 
    design <- model.matrix(~ pheno$Type + pheno$PID)
    rownames(design) <- rownames(pheno)
    colnames(design) <- c("Intercept",
                          paste(group_col, collapse = "-"), 
                          as.character(unique(pheno$PID)[-1]))    
  }else{
    design <- model.matrix(~0 + pheno$Type)
    rownames(design) <- rownames(pheno)
    colnames(design) <- group_col
  }

  # show distribution
  edata <- as.matrix(exprs(dataset))
  exprSet <- edata  
  boxplot(exprSet)
  plotDensities(exprSet)
  
  # Normalization: TMM
  if(scale){
    require(edgeR)
    DGEList <- edgeR::DGEList(
                        counts = exprSet, 
                        group = pheno$Type) 
    exprSet_norm <- edgeR::calcNormFactors(DGEList, method = "TMM")
    plotMDS(exprSet_norm, col=as.numeric(pheno$Type))    
  }else{
    exprSet_norm <- exprSet
  }

  # linear fitting 
  limma_voom <- voom(exprSet_norm, design, plot = TRUE)
  fit <- lmFit(limma_voom, design)
  
  if(pair){
    # eBayes
    fit2 <- eBayes(fit)
    qqt(fit2$t, df = fit2$df.prior+fit2$df.residual, pch = 16, cex = 0.2)
    abline(0,1)
    
    # differential features
    diff_gene <- topTable(fit2, number = Inf, adjust.method = 'BH', coef = 2) %>%
      rownames_to_column("GeneID")     
  }else{
    # contrast group for unpaired test  
    group <- paste(group_col, collapse = "-")
    if(group%in%"ParacancerousTissue-PrimaryCancerTissue"){
      contrast <- makeContrasts(contrasts = "ParacancerousTissue-PrimaryCancerTissue",
                                levels    = design)
    }else if(group%in%"PrimaryCancerTissue-LiverMetastasis"){
      contrast <- makeContrasts(contrasts = "PrimaryCancerTissue-LiverMetastasis",
                                levels    = design)
    }
    print(contrast)
    # eBayes
    fit2 <- contrasts.fit(fit, contrast)
    fit2 <- eBayes(fit2)
    
    qqt(fit2$t, df = fit2$df.prior+fit2$df.residual, pch = 16, cex = 0.2)
    abline(0,1)
    
    # differential features
    diff_gene <- topTable(fit2, number = Inf, adjust.method = 'BH', coef = 1) %>%
      rownames_to_column("GeneID")    
  }
  
  # validate the enriched directory
  data.frame(edata)[rownames(data.frame(edata))%in%diff_gene$GeneID[1], , F] %>% 
    t() %>% data.frame() %>%
    setNames("Gene") %>%
    rownames_to_column("SampleID") %>%
    inner_join(pheno%>%rownames_to_column("SampleID"), by = "SampleID") -> a1
  print(wilcox.test(Gene~Type, data = a1, paired=T))
  print(ggplot(a1, aes(x=Type, y=Gene))+
    geom_boxplot()+
    labs(y=diff_gene$GeneID[1], x="")+
    geom_line(aes(group=PID))+
    theme_bw())  
  
  # enriched directory
  diff_gene[which(diff_gene$logFC >= fc & diff_gene$adj.P.Val < Pval), "Enrichment"] <- group_col[1]
  diff_gene[which(diff_gene$logFC <= -fc & diff_gene$adj.P.Val < Pval), "Enrichment"] <- group_col[2]
  diff_gene[which(abs(diff_gene$logFC) < fc | diff_gene$adj.P.Val >= Pval), "Enrichment"] <- "Nonsignif"
  
  diff_gene$Block <- paste(group_col, collapse = "_vs_")
  
  res <- diff_gene %>% dplyr::select(GeneID, Block, logFC, adj.P.Val, Enrichment, everything()) %>%
    arrange(adj.P.Val, logFC) 
  
  print(dim(res %>% filter(Enrichment != "Nonsignif")))
  
  return(res)
}
```


### acquiring the ExpressionSet object via the group parameters
```{r}
ExprSet1 <- get_paired_data(ExprSet = genExprSet, 
                     group_col = c("ParacancerousTissue", "PrimaryCancerTissue"))
if(!file.exists("../../Result/profile/Paracancerous_PrimaryCancer_GeneExprSet.RDS")){
  saveRDS(ExprSet1, "../../Result/profile/Paracancerous_PrimaryCancer_GeneExprSet.RDS", compress = TRUE)
}

ExprSet2 <- get_paired_data(ExprSet = genExprSet, 
                     group_col = c("PrimaryCancerTissue", "LiverMetastasis"))
if(!file.exists("../../Result/profile/PrimaryCancer_Liver_GeneExprSet.RDS")){
  saveRDS(ExprSet2, "../../Result/profile/PrimaryCancer_Liver_GeneExprSet.RDS", compress = TRUE)
}
```


### Differential Expression Analysis
```{r}
DEA_DESeq2_1 <- get_DiffGene_DESeq2(
                    dataset   = ExprSet1,
                    group_col = c("ParacancerousTissue", "PrimaryCancerTissue"),
                    pair      = TRUE,
                    fc        = 1,
                    Pval      = 0.05)

if(!dir.exists("../../Result/Differential")){
  dir.create("../../Result/Differential")
}
write.csv(DEA_DESeq2_1$results, "../../Result/Differential/Paracancerous_PrimaryCancer_DESeq2_Gene.csv", 
          row.names = F)


DEA_DESeq2_2 <- get_DiffGene_DESeq2(
                    dataset   = ExprSet2,
                    group_col = c("PrimaryCancerTissue", "LiverMetastasis"),
                    pair      = TRUE,
                    fc        = 1,
                    Pval      = 0.05)
write.csv(DEA_DESeq2_2$results, "../../Result/Differential/PrimaryCancer_Liver_DESeq2_Gene.csv", 
          row.names = F)

DEA_limma_1 <- get_DiffGene_limma(
                    dataset   = ExprSet1,
                    group_col = c("ParacancerousTissue", "PrimaryCancerTissue"),
                    pair      = TRUE,
                    scale     = TRUE,
                    fc        = 1,
                    Pval      = 0.05)
write.csv(DEA_limma_1, "../../Result/Differential/Paracancerous_PrimaryCancer_limma_Gene.csv", 
          row.names = F)

DEA_limma_2 <- get_DiffGene_limma(
                    dataset   = ExprSet2,
                    group_col = c("PrimaryCancerTissue", "LiverMetastasis"),
                    pair      = TRUE,
                    scale     = TRUE,                    
                    fc        = 1,
                    Pval      = 0.05)
write.csv(DEA_limma_2, "../../Result/Differential/PrimaryCancer_Liver_limma_Gene.csv", 
          row.names = F)
```


### systemic information
```{r}
sessionInfo()
```


### Reference

1. [RNA-seq摸索：4. edgeR/limma/DESeq2差异基因分析](https://www.jianshu.com/p/bdf2b72b8761)

2. [DESeq2 - paired sample, interaction and fold change
](https://grokbase.com/t/r/bioconductor/1464bjd46r/bioc-deseq2-paired-sample-interaction-and-fold-change)

3. [DESeq2: Multi-factor designs](http://seqanswers.com/forums/archive/index.php/t-34614.html)

4. [Limma: Paired samples, multiple groups: problems understanding contrasts and model.matrix](https://support.bioconductor.org/p/94460/)

5. [limma paired t test](https://www.biostars.org/p/396703/)

