---
title: "Differential Gene Expression: Volcano plot"
date: "Created: 2021-03-22 Updated: `r Sys.Date()`"
output: 
  html_notebook:
    codes: hide
---

## Introduction

To illuminate the Differential Expressed Genes with parameters: |log2FC| > 1 and P.value < 0.05 among different groups, we used the following plots: **volcano plot** displays the DEG's names between diseases controls


```{r setup, include=FALSE}
knitr::opts_chunk$set(warning = F, message = F)
library(dplyr)
library(tibble)
library(ggplot2)
library(ggrepel)

options(stringsAsFactors = F)
options(future.globals.maxSize = 1000 * 1024^2)

mus.grp <- c("TAC_NC", "TAC", "HF_NC", "HF")
mus.grp.col <- c("#6C326C", "#77A2D1", "#FFD169", "#635F5F")

hsp.grp <- c("NC", "DCM", "HCM", "ICM", "NICM")
hsp.grp.col <- c("#6C326C", "#77A2D1", "#FFD169", "#635F5F", "#D4D09A")
```


### load data

1. Human: the number of samples (n=197)

  * NC:  69
  
  * DCM: 63
  
  * ICM: 65
  
  * HCM: 8
  
  * NICM: 20 


2. Mouse: the number of samples (n=85)

  * TAC_NC: 20 
    
  * TAC: 22 
    
  * HF_NC: 18
  
  * HF: 25
  
```{r}
hsp_DCM <- read.csv("../../Result/Differential/final/Hsp_DCM_NC_Diffgene.csv")
hsp_ICM <- read.csv("../../Result/Differential/final/Hsp_ICM_NC_Diffgene.csv")

mus_TAC <- read.csv("../../Result/Differential/final/Mus_TAC_NC_Diffgene.csv")
mus_HF <- read.csv("../../Result/Differential/final/Mus_HF_NC_Diffgene.csv")
```


### Differential gene expression 

* log2FoldChange: 1

* Adjust Pvalue: 0.05

* Plot cutoff: log2FoldChange either 2 or 4

```{r}
volcanofun <- function(datset=mus_TAC,
                       group_name=c("TAC_NC", "TAC"),
                       group_col=c("#6C326C", "#77A2D1"),
                       pval=0.05, 
                       fc=1,
                       fc2=3){
  
  # datset=mus_TAC
  # group_name=c("TAC_NC", "TAC")
  # group_col=c("#6C326C", "#77A2D1")
  # pval=0.01
  # fc=1
  # fc2=3
  
  dat <- datset %>% 
    mutate(color = case_when(logFC > fc & adj.P.Val < pval ~ group_name[2],
                             logFC < -fc & adj.P.Val < pval ~ group_name[1],
                             adj.P.Val > pval | abs(logFC) < fc ~ "Nonsignif")) %>%
    mutate(color=factor(color, levels = c(group_name, "Nonsignif")))
  # print(table(dat$color))
  dat_status <- table(dat$color)
  dat_status_number <- as.numeric(dat_status)
  dat_status_name <- names(dat_status)
  legend_label <- c(paste0(dat_status_name[1], " (", dat_status_number[1], ")"),
                    paste0(dat_status_name[2], " (", dat_status_number[2], ")"))
  
  dat.signif <- subset(dat, adj.P.Val < pval & abs(logFC) > fc2)
  
  print(table(dat.signif$color))
  
  group_col_new <- c(group_col, "darkgray")
  group_name_new <- levels(dat$color)
  
  xlabel <- paste0("log2(", paste(group_name, collapse="/"), ")")
  
  # Make a basic ggplot2 object with x-y values
  pl <- ggplot(dat, aes(x = logFC, y = -log10(adj.P.Val), color = color))+ 
          geom_point(size = 0.8, alpha = 0.8)+
          scale_color_manual(name = "Enrichment",
                             values = group_col_new,
                             labels = c(legend_label, "Nonsignif"))+
          xlab(xlabel) + 
          ylab(expression(-log[10]("adjusted p-value")))+ 
          geom_hline(yintercept=-log10(pval), alpha=.8, linetype=2, size=.7)+
          geom_vline(xintercept=fc, alpha=.8, linetype=2, size=.7)+
          geom_vline(xintercept=-fc, alpha=.8, linetype=2, size=.7)+ 
          geom_text_repel(data = dat.signif,
                          aes(label = GeneID),
                          size = 4,
                          max.overlaps = getOption("ggrepel.max.overlaps", default = 80),
                          segment.linetype = 1,
                          segment.curvature = -1e-20,
                          box.padding = unit(0.35, "lines"),
                          point.padding = unit(0.3, "lines"),
                          arrow = arrow(length = unit(0.005, "npc")),
                          color = "black",     # text color
                          bg.color = "white", # shadow color
                          bg.r = 0.15)+
          annotate("text", x=min(dat$logFC), y=-log10(pval), label=pval, size=6, color="red")+
          annotate("text", x=fc, y=0, label=fc, size=6, color="red")+
          annotate("text", x=-fc, y=0, label=-fc, size=6, color="red")+
          scale_y_continuous(trans = "log1p")+
          theme_bw()+ 
          theme(axis.title = element_text(face = 'bold',color = 'black',size = 14),
                axis.text = element_text(color = 'black',size = 10),
                text = element_text(size = 8, color = "black", family="serif"),
                panel.grid = element_blank(),
                legend.position = 'right',
                legend.key.height = unit(0.6,'cm'),
                legend.text = element_text(face = "bold", color = 'black',size = 10),
                strip.text = element_text(face = "bold", size = 14))
  return(pl)
}
```


### Human 

* "NC-DCM"
```{r, fig.width=6, fig.height=4}
DCM_pl <- volcanofun(datset = hsp_DCM, group_name = hsp.grp[1:2], 
                     group_col = hsp.grp.col[1:2], fc2 = 3)
DCM_pl
ggsave(filename = "../../Result/Figures/version_210322/SFig2A.pdf", DCM_pl, width = 6, height = 4)
```


* "NC-ICM"
```{r, fig.width=6, fig.height=4}
ICM_pl <- volcanofun(datset = hsp_ICM, group_name = hsp.grp[c(1, 4)], 
                     group_col = hsp.grp.col[c(1, 4)], fc2 = 3)
ICM_pl
ggsave(filename = "../../Result/Figures/version_210322/SFig2B.pdf", ICM_pl, width = 6, height = 4)
```


### Mouse 

* "TAC_NC-TAC"
```{r, fig.width=6, fig.height=4}
Mus_TAC_pl <- volcanofun(datset = mus_TAC, group_name = mus.grp[1:2], 
                         group_col = mus.grp.col[1:2], fc2 = 3)
Mus_TAC_pl
ggsave(filename = "../../Result/Figures/version_210322/SFig2C.pdf", Mus_TAC_pl, width = 6, height = 4)
```

* "HF_NC-HF"
```{r, fig.width=6, fig.height=4}
Mus_HF_pl <- volcanofun(datset = mus_HF, group_name = mus.grp[3:4], 
                        group_col = mus.grp.col[3:4], fc2 = 3)
Mus_HF_pl
ggsave(filename = "../../Result/Figures/version_210322/SFig2D.pdf", Mus_HF_pl, width = 6, height = 4)
```


### Systemic information
```{r}
sessionInfo()
```


### Reference 

1. [R包limma作差异基因分析](https://cloud.tencent.com/developer/article/1667505)
